### PR TITLE
SPARKC-163 Replace CassandraRelation by CassandraSourceRelation

### PR DESCRIPTION
**Pull request by alexliu68**

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala [hunk 1]
```diff
@@ -1,65 +1,157 @@
...
+        Seq.empty
+      }
+    } else {
+      val tables = mutable.Set[(String, Boolean)]()
+      for (ksDef <- schema.keyspaces) {
+        tables ++= ksDef.tables.map(
+          table => (s"${ksDef.keyspaceName}.${table.tableName}", false))
+      }
+      tables.toSeq
+    }
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala]: Why not simpler:
```scala
    val schema = Schema.fromCassandra(getCassandraConnector(tableRef), databaseName)
    for {
      ksDef <- schema.keyspaces.toSeq
      tableDef <- ksDef.tables
    } yield (s"${ksDef.keyspaceName}.${tableDef.tableName}", false) 
```

Schema.fromCassandra can fetch schema partially, only for the given keyspace.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala [hunk 2]
```diff
@@ -1,65 +1,157 @@
...
+    val clusterName = cluster.getOrElse(csc.getCluster)
+    val names = ListBuffer[String]()
+    while (iter.hasNext) {
+      val ident = iter.next()
+      val c = ident(0)
+      val db = ident(1)
+      val table = ident(2)
+      if (clusterName == c) {
+        if (databaseName.nonEmpty && databaseName.get == db) {
+          names += table
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala]: Why in this case it is not qualified by the database name? Is this distinction really needed?
The code for getTableForCassandra did not make this distinction and qualified all table names with a database name.

Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala]: Changed it to this:
```scala
  /** List all tables for a given database name and cluster from local cache */
  def getTablesFromCache(
      databaseName: Option[String],
      cluster: Option[String] = None): Seq[(String, Boolean)] = {

    import JavaConversions._
    val clusterName = cluster.getOrElse(csc.getCluster)
    for (Seq(c, db, table) <- cachedDataSourceTables.asMap().keySet().toSeq
         if c == clusterName && databaseName.forall(_ == db)) yield {
      (s"$db.$table", true)
    }
  }
```
and it passed all tests.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala [hunk 3]
```diff
@@ -1,65 +1,157 @@
...
+  override def unregisterAllTables(): Unit = {
+    cachedDataSourceTables.invalidateAll()
+  }
 
   override def tableExists(tableIdentifier: Seq[String]): Boolean = {
     val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
-    val schema = schemas.get(cluster)
-    val tabDef = for (ksDef <- schema.keyspaceByName.get(database); tabDef <- ksDef.tableByName.get(table)) yield tabDef
-    tabDef.isDefined
+    val tableRef = TableRef(table, database, Option(cluster))
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala]: tableRef is used only on one branch of the condition below, you can move it down

**Comment by pkolaczk** [comment]
Overall looks good. I created a separate PR for code simplifications. Please squash commits.

**Comment by alexliu68** [comment]
Squash commits is done.

**Comment by alexliu68** [comment]
retest it please

**Comment by alexliu68** [comment]
failed tests are not related to this ticket.

09:26:52 [info] *** 2 TESTS FAILED ***
09:26:52 [error] Failed: Total 295, Failed 2, Errors 0, Passed 293, Ignored 10
09:26:52 [error] Failed tests:
09:26:52 [error] 	com.datastax.spark.connector.util.MultiThreadedSpec
09:26:52 [error] 	com.datastax.spark.connector.streaming.ActorStreamingSpec
09:26:57  WARN 16:26:57,402 org.apache.cassandra.utils.CLibr

**Comment by pkolaczk** [comment]
+1

**Pull request merged**